integ-testing/src/main/java/com/google/firebase/testing/integ/StrictModeRule.java

```diff
@@ -0,0 +1,166 @@
+ * under test to run in the Main thread in production, please use the provided {@link
+ * #runOnMainThread(MaybeThrowingRunnable)} to execute it on the Main thread.
+ *
+ * <p>Example use:
+ *
+ * <pre>{@code
+ * @Test
+ * public class MyTests {
+ *   @Rule public StrictModeRule strictMode = new StrictModeRule();
+ *
+ *  @Test public void myTest() {
+ *    // runs on the instrumentation thread
+ *    runMyCode();
+ *
+ *    // runs on Main thread.
+ *    strictMode.runOnMainThread(() -> {
+ *      runCodeOnMainThread();
+ *    });
+ *  }
+ * }
+ * }</pre>
+ */
+public class StrictModeRule implements TestRule {
+
+  private static final Executor penaltyListenerExecutor = Runnable::run;
+
+  /** Runs {@code runnable} on Main thread. */
+  public <E extends Throwable> void runOnMainThread(MaybeThrowingRunnable<E> runnable) throws E {
+    try {
+      new UiThreadStatement(
+              new Statement() {
+                @Override
+                public void evaluate() throws E {
+                  runnable.run();
+                }
+              },
+              true)
+          .evaluate();
+    } catch (Throwable throwable) {
+      @SuppressWarnings("unchecked")
+      E e = (E) throwable;
+      throw e;
+    }
+  }
+
+  /** Runs {@code callable} on Main thread and returns it result. */
+  public <T, E extends Throwable> T runOnMainThread(MaybeThrowingCallable<T, E> callable) throws E {
+    try {
+      AtomicReference<T> result = new AtomicReference<>();
+      new UiThreadStatement(
+              new Statement() {
+                @Override
+                public void evaluate() throws E {
+                  result.set(callable.call());
+                }
+              },
+              true)
+          .evaluate();
+      return result.get();
+    } catch (Throwable throwable) {
+      @SuppressWarnings("unchecked")
+      E e = (E) throwable;
+      throw e;
+    }
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        AtomicReference<ThreadPolicy> originalThreadPolicy = new AtomicReference<>();
+        VmPolicy originalVmPolicy = StrictMode.getVmPolicy();
+
+        ConcurrentLinkedQueue<Throwable> violations = new ConcurrentLinkedQueue<>();
+
+        InstrumentationRegistry.getInstrumentation()
+            .runOnMainSync(
+                () -> {
+                  originalThreadPolicy.set(StrictMode.getThreadPolicy());
+
+                  StrictMode.setThreadPolicy(createThreadPolicy(violations));
+                  StrictMode.setVmPolicy(createVmPolicy(violations));
+                });
+        try {
+          base.evaluate();
+        } catch (Throwable e) {
+          violations.add(e);
+        } finally {
+          InstrumentationRegistry.getInstrumentation()
+              .runOnMainSync(() -> StrictMode.setThreadPolicy(originalThreadPolicy.get()));
+          runGc();
```

Comment:
out of curiosity, why is this part needed?

Reply:
added a comment with an explanation.